server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -467,10 +488,27 @@ public void uploadMetadata(
             );
             try {
                 IndexOutput indexOutput = storeDirectory.createOutput(metadataFilename, IOContext.DEFAULT);
+                List<SegmentCommitInfo> infos = segmentInfosSnapshot.asList();
+                Map<String, Version> segmentToLuceneVersion = new HashMap<>();
+                for (SegmentCommitInfo segmentCommitInfo : infos) {
+                    SegmentInfo info = segmentCommitInfo.info;
+                    Set<String> segFiles = info.files();
+                    for (String file : segFiles) {
+                        segmentToLuceneVersion.put(file, info.getVersion());
+                    }
+                }
+
                 Map<String, String> uploadedSegments = new HashMap<>();
                 for (String file : segmentFiles) {
                     if (segmentsUploadedToRemoteStore.containsKey(file)) {
-                        uploadedSegments.put(file, segmentsUploadedToRemoteStore.get(file).toString());
+                        UploadedSegmentMetadata metadata = segmentsUploadedToRemoteStore.get(file);
+                        if (segmentToLuceneVersion.containsKey(metadata.originalFilename)) {
+                            metadata.setWrittenBy(segmentToLuceneVersion.get(metadata.originalFilename));
+                        } else if (metadata.originalFilename.equals(segmentInfosSnapshot.getSegmentsFileName())) {
+                            metadata.setWrittenBy(segmentInfosSnapshot.getCommitLuceneVersion());
+                        }
```

Comment:
As `else` branch is missing, it is possible that `writtenBy` is not set in few cases. If it is to not happen at all, can we throw exception in `else`?

Reply:
Good point, will add! 
Would a `CorruptIndexException`, detailing the specific segment file name and that the Lucene version should not be missing, be the right type to throw in this case?

Reply:
Hey Sachin,
Resolving this. Please reopen if any changes are needed here.
Thanks!